src/plugins/dashboard/public/dashboard_api/panels_manager.ts

```diff
@@ -0,0 +1,465 @@
+  incomingEmbeddable: EmbeddablePackageState | undefined,
+  initialPanels: DashboardPanelMap,
+  initialPanelsRuntimeState: UnsavedPanelState,
+  trackPanel: ReturnType<typeof initializeTrackPanel>,
+  getReferencesForPanelId: (id: string) => Reference[],
+  pushReferences: (references: Reference[]) => void
+) {
+  const children$ = new BehaviorSubject<{
+    [key: string]: unknown;
+  }>({});
+  const panels$ = new BehaviorSubject(initialPanels);
+  function setPanels(panels: DashboardPanelMap) {
+    if (panels !== panels$.value) panels$.next(panels);
+  }
+  let restoredRuntimeState: UnsavedPanelState = initialPanelsRuntimeState;
+
+  function setRuntimeStateForChild(childId: string, state: object) {
+    restoredRuntimeState[childId] = state;
+  }
+
+  // --------------------------------------------------------------------------------------
+  // Place the incoming embeddable if there is one
+  // --------------------------------------------------------------------------------------
+  if (incomingEmbeddable) {
+    let incomingEmbeddablePanelState: DashboardPanelState;
+    if (
+      incomingEmbeddable.embeddableId &&
+      Boolean(panels$.value[incomingEmbeddable.embeddableId])
+    ) {
+      // this embeddable already exists, just update the explicit input.
+      incomingEmbeddablePanelState = panels$.value[incomingEmbeddable.embeddableId];
+      const sameType = incomingEmbeddablePanelState.type === incomingEmbeddable.type;
+
+      incomingEmbeddablePanelState.type = incomingEmbeddable.type;
+      setRuntimeStateForChild(incomingEmbeddable.embeddableId, {
+        // if the incoming panel is the same type as what was there before we can safely spread the old panel's explicit input
+        ...(sameType ? incomingEmbeddablePanelState.explicitInput : {}),
+
+        ...incomingEmbeddable.input,
+        id: incomingEmbeddable.embeddableId,
+
+        // maintain hide panel titles setting.
+        hidePanelTitles: incomingEmbeddablePanelState.explicitInput.hidePanelTitles,
+      });
+      incomingEmbeddablePanelState.explicitInput = {
+        id: incomingEmbeddablePanelState.explicitInput.id,
+      };
+    } else {
+      // otherwise this incoming embeddable is brand new.
+      const embeddableId = incomingEmbeddable.embeddableId ?? v4();
+      setRuntimeStateForChild(embeddableId, incomingEmbeddable.input);
+      const { newPanelPlacement } = runPanelPlacementStrategy(
+        PanelPlacementStrategy.findTopLeftMostOpenSpace,
+        {
+          width: incomingEmbeddable.size?.width ?? DEFAULT_PANEL_WIDTH,
+          height: incomingEmbeddable.size?.height ?? DEFAULT_PANEL_HEIGHT,
+          currentPanels: panels$.value,
+        }
+      );
+      incomingEmbeddablePanelState = {
+        explicitInput: { id: embeddableId },
+        type: incomingEmbeddable.type,
+        gridData: {
+          ...newPanelPlacement,
+          i: embeddableId,
+        },
+      };
+    }
+
+    setPanels({
```

Comment:
`initializePanelsManager` is guaranteed to run before anyone has subscribed to panels subject. 

src/plugins/dashboard/public/dashboard_container/external_api/dashboard_renderer.tsx awaits promise from `loadDashboardApi` before calling `setDashboardApi`. `loadDashboardApi` does not return until `getDashboardApi` returns. `getDashboardApi` does not return until `initializePanelsManager` returns.

Reply:
It seems like a code-smell to me to initialize a behaviour subject then synchronously call `.next` on it instead of creating the behaviour subject with the initial value. That said, it's okay to keep it as you're certain that no subscriptions are listening to that first `next` call.

Reply:
Added this thread to https://github.com/elastic/kibana/issues/203130 so both can be resolved at the same time since they are in the same area.